javascript/ql/src/semmle/javascript/NullSensitiveContext.qll

```diff
@@ -0,0 +1,488 @@
+}
+
+/*
+ * _ >>> y
+ */
+
+class URShiftExprLeftOperand extends NullSensitiveContext, URShiftExpr {
+  override Expr getPrincipalArgument() { result = this.getLeftOperand() }
+}
+
+/*
+ * x >>> _
+ */
+
+class URShiftExprRightOperand extends NullSensitiveContext, URShiftExpr {
+  override Expr getPrincipalArgument() { result = this.getRightOperand() }
+}
+
+/*
+ * ~_
+ */
+
+class BitNotExprOperand extends NullSensitiveContext, BitNotExpr {
+  override Expr getPrincipalArgument() { result = this.getOperand() }
+}
+
+/*
+ * _ &= y
+ */
+
+class AssignAndExprLhs extends NullSensitiveContext, AssignAndExpr {
+  override Expr getPrincipalArgument() { result = this.getLhs() }
+}
+
+/*
+ * x &= _
+ */
+
+class AssignAndExprRhs extends NullSensitiveContext, AssignAndExpr {
+  override Expr getPrincipalArgument() { result = this.getRhs() }
+}
+
+/*
+ * _ |= y
+ */
+
+class AssignOrExprLhs extends NullSensitiveContext, AssignOrExpr {
+  override Expr getPrincipalArgument() { result = this.getLhs() }
+}
+
+/*
+ * x |= _
+ */
+
+class AssignOrExprRhs extends NullSensitiveContext, AssignOrExpr {
+  override Expr getPrincipalArgument() { result = this.getRhs() }
+}
+
+/*
+ * _ ^= y
+ */
+
+class AssignXOrExprLhs extends NullSensitiveContext, AssignXOrExpr {
+  override Expr getPrincipalArgument() { result = this.getLhs() }
+}
+
+/*
+ * x ^= _
+ */
+
+class AssignXOrExprRhs extends NullSensitiveContext, AssignXOrExpr {
+  override Expr getPrincipalArgument() { result = this.getRhs() }
+}
+
+/*
+ * _ <<= y
+ */
+
+class AssignLShiftExprLhs extends NullSensitiveContext, AssignLShiftExpr {
+  override Expr getPrincipalArgument() { result = this.getLhs() }
+}
+
+/*
+ * x <<= _
+ */
+
+class AssignLShiftExprRhs extends NullSensitiveContext, AssignLShiftExpr {
+  override Expr getPrincipalArgument() { result = this.getRhs() }
+}
+
+/*
+ * _ >>= y
+ */
+
+class AssignRShiftExprLhs extends NullSensitiveContext, AssignRShiftExpr {
+  override Expr getPrincipalArgument() { result = this.getLhs() }
+}
+
+/*
+ * x >>= _
+ */
+
+class AssignRShiftExprRhs extends NullSensitiveContext, AssignRShiftExpr {
+  override Expr getPrincipalArgument() { result = this.getRhs() }
+}
+
+/*
+ * _ >>>= y
+ */
+
+class AssignURShiftExprLhs extends NullSensitiveContext, AssignURShiftExpr {
+  override Expr getPrincipalArgument() { result = this.getLhs() }
+}
+
+/*
+ * x >>>= _
+ */
+
+class AssignURShiftExprRhs extends NullSensitiveContext, AssignURShiftExpr {
+  override Expr getPrincipalArgument() { result = this.getRhs() }
+}
+
+/*
+ * pat = _
+ */
+
+class DestructuringAssignExprRhs extends NullSensitiveContext, AssignExpr {
+  DestructuringAssignExprRhs() { this.getLhs() instanceof DestructuringPattern }
+
+  override Expr getPrincipalArgument() { result = this.getRhs() }
+}
+
+/*
+ * [..._]
+ */
+
+class SpreadElementOperand extends NullSensitiveContext, SpreadElement {
+  override Expr getPrincipalArgument() { result = this.getOperand() }
+}
```

Comment:
You may also want to add a subclass for the iteration domain of `for`-`of` loops (`for`-`in` loops, on the other hand, are fine).